File: Source/AlphaTab/Rendering/Glyphs/NoteVibratoGlyph.cs
using System;
using AlphaTab.Model;
using AlphaTab.Platform;

namespace AlphaTab.Rendering.Glyphs
{
    class NoteVibratoGlyph : GroupedEffectGlyph
    {
        public const int SlightWaveOffset = 10;
        public const float SlightWaveSize = 8.5f;

        private readonly VibratoType _type;
        private readonly float _scale;
        private MusicFontSymbol _symbol;
        private float _symbolSize;
        private float _symbolOffset;

        public NoteVibratoGlyph(float x, float y, VibratoType type, float scale = 1.2f)
            : base(BeatXPosition.EndBeat)
        {
            _type = type;
            _scale = scale;
            X = x;
            Y = y;
        }

        public override void DoLayout()
        {
            base.DoLayout();

            var symbolHeight = 0f;
            switch (_type)
            {
                case VibratoType.Slight:
                    _symbol = MusicFontSymbol.WaveHorizontalSlight;
                    _symbolSize = SlightWaveSize * _scale;
                    _symbolOffset = SlightWaveOffset * _scale;
                    symbolHeight = 6 * _scale;
                    break;
                case VibratoType.Wide:
                    _symbol = MusicFontSymbol.WaveHorizontalWide;
                    _symbolSize = 10 * _scale;
                    _symbolOffset = 7 * _scale;
                    symbolHeight = 10 * _scale;
                    break;
            }

            Height = symbolHeight * Scale;
        }

        protected override void PaintGrouped(float cx, float cy, float endX, ICanvas canvas)
        {
            var startX = cx + X;
            var width = endX - startX;
            var step = _symbolSize * Scale;
            var loops = (int)Math.Max(1, width / step);

            var loopX = 0f;
            for (var i = 0; i < loops; i++)
            {
                canvas.FillMusicFontSymbol(cx + X + loopX, cy + Y + _symbolOffset, _scale, _symbol);
                loopX += step;
            }
        }
    }
}

File: src/rendering/glyphs/NoteVibratoGlyph.ts
import { VibratoType } from '@src/model/VibratoType';
import type { ICanvas } from '@src/platform/ICanvas';
import { BeatXPosition } from '@src/rendering/BeatXPosition';
import { GroupedEffectGlyph } from '@src/rendering/glyphs/GroupedEffectGlyph';
import { MusicFontSymbol } from '@src/model/MusicFontSymbol';

export class NoteVibratoGlyph extends GroupedEffectGlyph {
    private _type: VibratoType;
    private _symbol: MusicFontSymbol = MusicFontSymbol.None;
    private _symbolSize: number = 0;
    private _symbolOffsetY: number = 0;
    private _partialWaves: boolean;

    public constructor(x: number, y: number, type: VibratoType, partialWaves: boolean = false) {
        super(BeatXPosition.EndBeat);
        this._type = type;
        this.x = x;
        this.y = y;
        this._partialWaves = partialWaves;
    }

    public override doLayout(): void {
        super.doLayout();
        switch (this._type) {
            case VibratoType.Slight:
                this._symbol = MusicFontSymbol.WiggleTrill;
                break;
            case VibratoType.Wide:
                this._symbol = MusicFontSymbol.WiggleVibratoMediumFast;
                break;
        }

        this._symbolSize = this.renderer.smuflMetrics.glyphWidths.get(this._symbol)!;
        this._symbolOffsetY = this.renderer.smuflMetrics.glyphTop.get(this._symbol)!;
        this.height = this.renderer.smuflMetrics.glyphHeights.get(this._symbol)!;
    }

    protected paintGrouped(cx: number, cy: number, endX: number, canvas: ICanvas): void {
        const startX: number = cx + this.x;
        const width: number = endX - startX;
        const step: number = this._symbolSize;

        // TODO: respect overlap for calculation?
        let loops: number = width / step;
        if (!this._partialWaves) {
            loops = Math.floor(loops);
        }
        if (loops < 1) {
            loops = 1;
        }

        const symbols:MusicFontSymbol[]=[];
        for (let i: number = 0; i < loops; i++) {
            symbols.push(this._symbol);
        }

        canvas.fillMusicFontSymbols(
            cx + this.x,
            cy + this.y + this._symbolOffsetY,
            1,
            symbols,
            false
        );
    }
}
